lib/Doctrine/ORM/Cache/MultiGetCollectionHydrator.php

```diff
@@ -0,0 +1,110 @@
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * This software consists of voluntary contributions made by many individuals
+ * and is licensed under the MIT license. For more information, see
+ * <http://www.doctrine-project.org>.
+ */
+
+namespace Doctrine\ORM\Cache;
+
+use Doctrine\ORM\Query;
+use Doctrine\ORM\PersistentCollection;
+use Doctrine\ORM\Mapping\ClassMetadata;
+use Doctrine\ORM\EntityManagerInterface;
+
+/**
+ * Collection hydrator that expects a target region to be instance of {Doctrine\ORM\Cache\MultiGetRegion},
+ * to enable loading the entire collection with one cache request.
+ *
+ * @since   2.5
+ * @author  Asmir Mustafic <goetas@gmail.com>
+ */
+class MultiGetCollectionHydrator implements CollectionHydrator
+{
+    /**
+     * @var \Doctrine\ORM\EntityManagerInterface
+     */
+    private $em;
+
+    /**
+     * @var \Doctrine\ORM\UnitOfWork
+     */
+    private $uow;
+
+    /**
+     * @var MultiGetRegion
+     */
+    private $targetRegion;
+
+    /**
+     * @var array
+     */
+    private static $hints = array(Query::HINT_CACHE_ENABLED => true);
+
+    /**
+     * @param \Doctrine\ORM\EntityManagerInterface $em The entity manager.
+     */
+    public function __construct(EntityManagerInterface $em, MultiGetRegion $targetRegion)
+    {
+        $this->em  = $em;
+        $this->uow = $em->getUnitOfWork();
+        $this->targetRegion = $targetRegion;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function buildCacheEntry(ClassMetadata $metadata, CollectionCacheKey $key, $collection)
+    {
+        $data = array();
+
+        foreach ($collection as $index => $entity) {
+            $data[$index] = $this->uow->getEntityIdentifier($entity);
+        }
+
+        return new CollectionCacheEntry($data);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function loadCacheEntry(ClassMetadata $metadata, CollectionCacheKey $key, CollectionCacheEntry $entry, PersistentCollection $collection)
+    {
+        $assoc           = $metadata->associationMappings[$key->association];
+        $list            = array();
+
+        $keys = array();
+        foreach ($entry->identifiers as $index => $identifier) {
+            $keys[$index] = new EntityCacheKey($assoc['targetEntity'], $identifier);
+        }
+
+        $entityEntries = $this->targetRegion->getMulti($keys);
```

Comment:
I'd inject the `CollectionCacheEntry` straight into the `getMulti()` method. This would simplify next loop which could benefit of consuming Collection API directly.